### PR TITLE
Change checked_add/sub/mul intrinsics to return overflow flag instead of error

### PR DESCRIPTION
**Pull request by simonbyrne**
This is to allow us to catch overflow in a more efficient manner. This is a first pass to implement the intrinsics: once this is settled I can go through and expose this in Julia itself.

This avoids the bootstrap issues of #15040.

### CODE REVIEW

**Comment by simonbyrne** [comment]
We also need a better name than `xchecked_add`: any suggestions?

File: base/checked.jl [hunk 1]
```diff
@@ -184,21 +195,32 @@ checked_add{T}(x1::T, x2::T, x3::T, x4::T, x5::T, x6::T, x7::T) =
...
 """
     Base.checked_sub(x, y)
 
 Calculates `x-y`, checking for overflow errors where applicable.
 
 The overflow protection may impose a perceptible performance penalty.
 """
-function checked_sub end
-
 function checked_sub{T<:SignedInt}(x::T, y::T)
```
Comment (tkelman) [inline base/checked.jl]: this makes the docstring only attached to the signed-specific signature

File: src/intrinsics.cpp [hunk 2]
```diff
@@ -1245,9 +1245,21 @@ static Value *emit_untyped_intrinsic(intrinsic f, Value *x, Value *y, Value *z,
 #else
         Value *res = builder.CreateCall2(intr, ix, iy);
 #endif
+        Value *val = builder.CreateExtractValue(res, ArrayRef<unsigned>(0));
         Value *obit = builder.CreateExtractValue(res, ArrayRef<unsigned>(1));
-        raise_exception_if(obit, prepare_global(jlovferr_var), ctx);
-        return builder.CreateExtractValue(res, ArrayRef<unsigned>(0));
+	Value *obyte = builder.CreateZExt(obit, T_int8);
```
Comment (tkelman) [inline src/intrinsics.cpp]: tabs

**Comment by pabloferz** [comment]
> We also need a better name than xchecked_add: any suggestions?

`_checked_add`?

**Comment by eschnett** [comment]
What about `add_overflows` or `carry_from_add`?

**Comment by vtjnash** [comment]
the llvm names (`*_with_overflow`) seems like a good choice to me, esp. since in the `uadd` case, overflow is actually the carry bit

File: src/intrinsics.cpp [hunk 3]
```diff
@@ -427,7 +427,7 @@ static jl_cgval_t generic_box(jl_value_t *targ, jl_value_t *x, jl_codectx_t *ctx
     if (jl_is_type_type(bt_value.typ))
         bt = jl_tparam0(bt_value.typ);
 
-    if (!bt || !jl_is_bitstype(bt)) {
+    if (!bt || !jl_isbits(bt)) {
```
Comment (vtjnash) [inline src/intrinsics.cpp]: don't change this test

File: src/intrinsics.cpp [hunk 4]
```diff
@@ -452,7 +452,7 @@ static jl_cgval_t generic_box(jl_value_t *targ, jl_value_t *x, jl_codectx_t *ctx
         || !jl_is_bitstype(v.typ)
         || jl_datatype_size(v.typ) != nb) {
         Value *typ = emit_typeof_boxed(v, ctx);
-        if (!jl_is_bitstype(v.typ)) {
+        if (!jl_isbits(v.typ)) {
```
Comment (vtjnash) [inline src/intrinsics.cpp]: or this test

File: base/float16.jl [hunk 5]
```diff
@@ -154,7 +154,7 @@ end
 
 for func in (:div,:fld,:cld,:rem,:mod,:atan2,:hypot)
     @eval begin
-        $func(a::Float16,b::Float16) = Float16($func(Float32(a),Float32(a)))
+        $func(a::Float16,b::Float16) = Float16($func(Float32(a),Float32(b)))
```
Comment (eschnett) [inline base/float16.jl]: That change is unrelated?

Comment (simonbyrne) [inline base/float16.jl]: Whoops, this leaked through from #17148

**Comment by simonbyrne** [comment]
Can anyone tell me why this is failing on everything except OS X?

**Comment by simonbyrne** [comment]
Ah, nevermind: it's because no Bool methods are defined.

**Comment by simonbyrne** [comment]
Do we need both `checked_neg` and the single-argument version of `checked_sub`?

cc: @eschnett

**Comment by eschnett** [comment]
@simonbyrne I'm not aware of a single-argument version of `checked_sub`. We can rename `checked_neg` to be a single-argument `checked_sub`, but I'd find that somewhat confusing. I would get rid of the single-argument `checked_sub`. I don't think it currently exists.

**Comment by simonbyrne** [comment]
Ah, it was only defined for `Bool`: I'll just disable it, and we can keep `checked_neg` for now.

I guess the main argument for single-argument `checked_sub` is that it matches the use of `-`.

**Comment by simonbyrne** [comment]
Okay, I'm happy with this now. Any comments/objections?

**Comment by StefanKarpinski** [comment]
The obvious: don't merge now since we're in feature freeze for 0.5.

**Comment by simonbyrne** [comment]
Now 0.5 has been released, can we get this merged? I need this for the rationals revamp.

File: base/checked.jl [hunk 6]
```diff
@@ -295,11 +307,8 @@ Calculates `div(x,y)`, checking for overflow errors where applicable.
 
 The overflow protection may impose a perceptible performance penalty.
 """
-function checked_div end
-
 # Base.div already checks; nothing to do here
```
Comment (tkelman) [inline base/checked.jl]: not sure if it matters, but there probably shouldn't be a comment between a docstring and the function it documents?

**Comment by tkelman** [comment]
let's check @nanosoldier `runbenchmarks(ALL, vs=":master")`

File: src/runtime_intrinsics.c [hunk 7]
```diff
@@ -501,10 +501,10 @@ cmp_iintrinsic(name, u)
 
 typedef int (*intrinsic_checked_t)(unsigned, void*, void*, void*);
 SELECTOR_FUNC(intrinsic_checked)
-#define checked_iintrinsic(name, u) \
+#define checked_iintrinsic(name, u, lambda_checked)				 \
```
Comment (tkelman) [inline src/runtime_intrinsics.c]: spaces instead of tabs

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-249465581) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/fabc781802d46be6758c9f1888cc02f105f768b5/b5165bd_vs_f7c8b29/report.md). cc @jrevels

File: src/runtime_intrinsics.c (hunk 7 above)
Comment (simonbyrne) [inline src/runtime_intrinsics.c]: done

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-250694163) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ac09477f71cddace301caeeba14e76372edfc017/6d3aae4_vs_c40090f/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-250737368) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/4350749688805ab93d59497d0fd793651df22894/f354c4d_vs_f935a50/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Okay, some basic inlining seems to have fixed the main regressions. I think the string-join one is spurious, as I can't recreate it locally

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-251378857) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/022bbd21a8c61f26e124ccb3198625ae06987135/41cc93a_vs_612c528/report.md). cc @jrevels

**Comment by KristofferC** [comment]
The simd one is there in both of the last two runs. I also don't remember it being noisy.

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-252210786) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1040b86d2c5a92617a3c729c63c3e309e56876d2/c2cf617_vs_38a63bd/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Okay, fixed the SIMD regressions with some strategic inlining. Will look into the subarray ones.

**Comment by simonbyrne** [comment]
@nanosoldier `runbenchmarks("array", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17152#issuecomment-252266433) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f8e4a54c8531001a9ada7326a96a56b5291250cb/c2cf617_vs_38a63bd/report.md). cc @jrevels

**Comment by simonbyrne** [comment]
Okay, they seem to have been spurious.

**Comment by JeffBezanson** [comment]
LGTM.

**Comment by JeffBezanson** [comment]
I think this needs some fallback definitions, or at least definitions for `BigInt`.

**Comment by tkelman** [comment]
Was this responsible for https://discourse.julialang.org/t/sorting-big-int-with-v-0-6/1241 ?
```
julia> sort([big(3), big(2)])
ERROR: MethodError: no method matching sub_with_overflow(::BigInt, ::BigInt)
```

**Comment by simonbyrne** [comment]
Yes, this and #18830.
